File: cam_node/src/cam.py
#!/usr/bin/env python3
import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2

def main():


    backends = [cv2.CAP_DSHOW, cv2.CAP_MSMF, cv2.CAP_ANY]

    
    # for backend in backends:
    #     for i in range(10):  
    #         cap = cv2.VideoCapture(i, backend)
    #         if cap.isOpened():
                
    #             name = cap.getBackendName()
    #             print("cam {} : backend {}. name: {}".format(i, backend, name))
    #             cap.release
    #     # Initialize the ROS node
    rospy.init_node('opencv_webcam_node')
    
    # Create a publisher to publish webcam images
    image_pub = rospy.Publisher('camera/image_raw', Image, queue_size=1)
    
    # Initialize the CvBridge
    bridge = CvBridge()
    
    # Open the default webcam using OpenCV's VideoCapture
    # On Windows, you might want to use cv2.CAP_DSHOW as the second parameter
    cap = cv2.VideoCapture(1,cv2.CAP_DSHOW)
    if not cap.isOpened():
        rospy.logerr("Failed to open webcam")
        return -1
    
    target_width = 1024
    target_height = 768
    
    rate = rospy.Rate(60)  # 30 Hz
    while not rospy.is_shutdown():
        # Capture a new frame from the webcam
        ret, frame = cap.read()
        if not ret:
            rospy.logwarn("Empty frame received")
            continue
        
        # Convert the OpenCV image (BGR format) to ROS Image message using cv_bridge
        frame_resized = cv2.resize(frame, (target_width, target_height))
        frame_flipped = cv2.flip(frame_resized, 0)
        msg = bridge.cv2_to_imgmsg(frame_flipped, encoding="bgr8")
        image_pub.publish(msg)
        
        rate.sleep()
    
    # Release the webcam when done
    cap.release()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: cam_node/src/cam.cpp
// opencv_webcam_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>
#include "opencv2/imgcodecs.hpp"

int main(int argc, char** argv) {
    // Initialize the ROS node
    ros::init(argc, argv, "opencv_webcam_node");
    ros::NodeHandle nh;
    
    // Create a publisher to publish webcam images
    ros::Publisher image_pub = nh.advertise<sensor_msgs::Image>("camera/image_raw", 1);

    // Open the default webcam using OpenCV's VideoCapture (DirectShow backend is used on Windows)
    cv::VideoCapture cap(0,cv::CAP_DSHOW);
    if (!cap.isOpened()) {
        ROS_ERROR("Failed to open webcam");
        return -1;
    }

    ros::Rate loop_rate(30); // 30 Hz
    while (ros::ok()) {
        cv::Mat frame;
        cap >> frame; // Capture a new frame from the webcam
        if (frame.empty()) {
            ROS_WARN("Empty frame received");
            continue;
        }
        
        // Convert the OpenCV image (BGR format) to ROS Image message using cv_bridge
        sensor_msgs::ImagePtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();
        image_pub.publish(msg);
        
        ros::spinOnce();
        loop_rate.sleep();
    }
    return 0;
}
